W-14001035-JAR-signature-validator-LDS (#702)
mulesoft/docs-runtime-manager: runtime-manager/modules/ROOT/pages/rtm-agent-jar-signature-validator.adoc

= JAR Signature Validator
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:page-deployment-options: hybrid

By default, the application validator service implements a JAR signature validator to prevent the deployment of non-authorized artifacts. 
Each signed entry in the artifact JAR file must match one of the keystore aliases. 

> You can indicate which single keystore alias you want to use to validate all artifacts deployed in the runtime. Artifacts signed with expired certificates are allowed.

Reviewer: ```suggestion
You can indicate which keystore alias to use to validate all artifacts deployed in the runtime. Artifacts signed with expired certificates are allowed.
```

== Configure the JAR Signature Validator

> Modify the file `$MULE_HOME/conf/mule-agent.yml` to include the following:

Reviewer: ```suggestion
Modify the `$MULE_HOME/conf/mule-agent.yml` file to include the following:
```

[source,java,linenums]
----
 services:
  mule.agent.artifact.validator.service:
    enabled: true
    validators:
    - type: jarSignatureValidator
      name: defaultJarSignatureValidator
      enabled: true
      args:
        truststore: signers-truststore.jks
        truststoreType: JKS
        truststoreAlias: code-signer-alias-1
        truststorePassword: '![PBEWITHSHA1ANDDESEDE,U/nA0wV714oNi9LHoAkDtiwg22ngG6b7yFTtg7fk2d4=]'
----

Use the following properties to configure the JAR signature validator:

[%header,cols="10a,60a"]
|===
|Key |Description

> |`type`| Required. `jarSignatureValidator` is the validator type implemented in the Runtime Manager agent plugin to verify the signature of all JAR files used for deployments in the Mule runtime.

Reviewer: ```suggestion
|`type`| Required. `jarSignatureValidator` is the validator type implemented in the Runtime Manager agent plugin to verify the signature of all JAR files used for deployments in Mule runtime.
```

|`name`| Required. A validator type can have multiple implementations. The name for the `jarSignatureValidator` type implemented in the Runtime Manager agent is `defaultJarSignatureValidator`.
|`enabled`| Optional. Set to `true` to enable the validator implementation or `false` to disable it. The default value is `true`.
|===

Use the following arguments to configure the JAR signature validator:

[%header,cols="10a,60a"]
|===
|Key |Description

> |`truststore`| Optional. The truststore file name that contains one or more certificates used to validate the signature of all JAR files that you use for deployments. The default location is `$MULE_HOME/conf`. You can use a relative path to reference the truststore file. If you don’t provide any, the validator loads the cacerts truststore file used by the JVM.

Reviewer: ```suggestion
|`truststore`| Optional. The truststore filename that contains one or more certificates used to validate the signature of all JAR files that you use for deployments. The default location is `$MULE_HOME/conf`. You can use a relative path to reference the truststore file. If you don’t provide a path, the validator loads the cacerts truststore file used by the JVM.
```

> |`truststoreType`| Optional. JKS or PKCS12. The default value is JKS.

Reviewer: ```suggestion
|`truststoreType`| Optional. `JKS` (default) or `PKCS12`.
```

> |`truststoreAlias`| Optional. Each certificate stored in a keystore file is identified by an alias string. This property specifies a single alias used to validate all JAR files. If you don't provide an alias, the validator uses all certificates in the keystore file to match the signature in the JAR files.

Reviewer: I don't know much about keystores and trustores - is the description supposed to say "keystore file" instead of "truststore"?

Author: Yes! I confirmed with the devs that if the name of the key is "trustore", the description should also refer to "trustore". I'll fix that. 🙂

> |`truststorePassword`| Optional for JKS, required for PKCS12. The keystore file password.

Reviewer: keystore > truststore?

|===

> == Start the Mule

Reviewer: ```suggestion
== Start Mule Runtime Engine
```

> Start the Mule and verify that the validator configuration was successful. The console shows the following logs:

Reviewer: ```suggestion
Start Mule runtime engine and verify that the validator configuration is successful. The console shows the following logs:
```
